src/main/java/com/limechain/trie/structure/NodeHandle.java

```diff
@@ -0,0 +1,115 @@
+package com.limechain.trie.structure;
+
+import com.limechain.trie.structure.nibble.Nibble;
+import com.limechain.trie.structure.nibble.Nibbles;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+/**
+ * Indicates that the given {@link Entry} is an actual node in the trie.
+ * <br>
+ * <br>
+ * This {@link NodeHandle} could be either one of:
+ * <ul>
+ *     <li>a {@link StorageNodeHandle}, meaning the node has a storage value (could be either leaf or branch node)</li>
+ *     <li>a {@link BranchNodeHandle}, meaning the node is a purely structural branch node with no storage value.</li>
+ * </ul>
+ */
+public abstract sealed class NodeHandle<T> extends Entry<T> permits StorageNodeHandle, BranchNodeHandle {
+    protected final int rawNodeIndex;
+
+    protected NodeHandle(TrieStructure<T> trieStructure, int rawNodeIndex) {
+        super(trieStructure);
+        this.rawNodeIndex = rawNodeIndex;
+    }
+
+    //NOTE:
+    // Later on, we might switch from optionals to nullable return types
+    // Depends on whether Optionals will be more helpful or more annoying
```

Comment:
This note applies also to TrieStructure's public interface. In general, currently we've got a mixture of both approaches: Optionals and nullable return types.

The original idea was:
- in places where "returning nothing" is a more valid and common use case, I've gone with Optional
- methods which are expected to be called when the user site is informed what needs to happen (i.e. a non-null value is expected to exists, missing values are rather an exception), I've gor for the @Nullable return type. A bit annoying is that this annotation enforces null-checks on user site, but I find that rather useful (we want to communicate to the user, that we know and it's intentional, that his value could be missing).

But what we ended up with is rather a mess, so I'm thinking of unifying the interface in favor of one of both alternatives.

Reply:
This should be a whole team discussion. The reason I like Optionals is because the Rust implementations have a lot of them and it's way easier to translate them into Java Optional. The reason I like `nulls` is because they are usually easier to write and the `@Nullable` annotation makes it very hard to forget to check a nullable field for null. I'm also not sure whether completely removing either option is a good solution. We can discuss this on tomorrow's daily for example.

Reply:
We left this thread intentionally unresolved for later traceability.